src/decision/src/GpsDecision.cpp

```diff
@@ -17,19 +19,133 @@ GpsDecision::GpsDecision(int argc, char **argv, std::string node_name) {
+    double current_x=0.0;
+    double current_y=0.0;
+    double current_z=0.0;
+
+    double x=relative_gps->x;
+    double y=relative_gps->y;
+    double z=relative_gps->z;
+
+    double distance=sqrt(pow((x-current_x),2)+pow((y-current_y),2)+pow((z-current_z),2));
+
+  return distance;
+}
+
+double GpsDecision::desiredAngle(const geometry_msgs::Point relative_gps,
+                                 float current_heading,geometry_msgs::Point currentPoint) {
+    const double TO_DEGREES=180/(M_PI);
+    const double TO_RADIANS=(M_PI)/180;
+
+    double next_x=relative_gps.x;
+    double next_y=relative_gps.y;
+    //calculate the relative x and y
+    double x=(next_x-currentPoint.x);
+    double y=(next_y-currentPoint.y);
+
+    //double dis=distance(relative_gps);
+    double tan=y/x;
+    double angle=(atan(tan));
+    double AngleRelativeNorth;
+    double desiredAngle;
+    AngleRelativeNorth=angle*TO_DEGREES;
+    if(x>=0 && y>=0) { //in the 1st quadrant
+        AngleRelativeNorth=90.0-AngleRelativeNorth;
+    }
+    else if (x<0 && y<0) { //in the 3rd quadrant
+        AngleRelativeNorth=-(90+AngleRelativeNorth);
+    }
+    else if (x>=0 && y<0) { //In the 4th quadrant
+        AngleRelativeNorth=90-AngleRelativeNorth;
+    }
+    else if(x<=0 && y==0) {
+        AngleRelativeNorth=-90;
+    }
+    else if(x==0 && y<0) {
+        AngleRelativeNorth=180;
+    }
+
+    //calcute the angle relative to current heading
+     if(current_heading>AngleRelativeNorth) {
+         if(current_heading-AngleRelativeNorth<=180) {
+             desiredAngle=-(current_heading-AngleRelativeNorth);
+         }
+         else {
+             desiredAngle=360-(current_heading-AngleRelativeNorth);
+         }
+     }
+     else if(current_heading<=AngleRelativeNorth) {
+
+         desiredAngle=AngleRelativeNorth-current_heading;
+         if(desiredAngle>180) {
+             desiredAngle=-(360-desiredAngle);
+         }
+     }
+    if(x==0 && y==0 && current_heading==0) {
+        desiredAngle==0;
+    }
+    desiredAngle=desiredAngle*TO_RADIANS;
+    return desiredAngle;
+}
+
+
+/**this function is used for rotatation */
+
+void  GpsDecision::rotate(double desiredAngle,double angular_velocity) {
+    geometry_msgs::Twist vel_msg;
+
+    vel_msg.angular.x=0;
+    vel_msg.angular.y=0;
+
+    if(desiredAngle>0){  //need to turn clockwise
+        vel_msg.angular.z=-abs(desiredAngle);
+    }
+    else if(desiredAngle<0) { //need to turn counter-clockwise
+        vel_msg.angular.z=abs(desiredAngle);
+    }
+
+    double current_angle;
+    double t0=ros::Time::now().toSec();
+    ros::Rate loop_rate(1000);
+    do{
+        //velocity_publisher.publish(vel_msg);
+        double t1=ros::Time::now().toSec();
+        current_angle=angular_velocity*(t1-t0);
+        ros::spinOnce();
+        loop_rate.sleep();
+        //loop_rate
+    }while(current_angle<desiredAngle);
+    vel_msg.angular.z=0;
+    // velocity_publisher.publish(vel_msg);
```

Comment:
Commented out code should be removed. That being said, I think you need this line to publish the twist message you create.